File: bézier/julia/04-bezier-matrix-straight.jl
# ============================================================
# Bezier Curve — Matrix Method  [P(u) = U · M · G]
# Julia port of 04-bezier-matrix-straight.py
# No external dependencies — plain string table formatting
# ============================================================

using Printf


# ============================================================
# -- Switch  (only change this one line) ---------------------
# ============================================================

const CASE = 2   # 1 = Quadratic 2D,  2 = Quartic 3D


# ============================================================
# -- Case Data  (do not change) ------------------------------
# ============================================================

const CASE_DATA = Dict(
    1 => (
        points   = Float64[1 1;
                           3 3;
                           5 1],
        u_values = [0.00, 0.25, 0.50, 0.75, 1.00],
        axes     = ["x", "y"],
    ),
    2 => (
        points   = Float64[0 0 1;
                           0 4 1;
                           4 0 1;
                           4 4 1;
                           5 4 1],
        u_values = [0.00, 0.25, 0.50, 0.75, 1.00],
        axes     = ["x", "y", "z"],
    ),
)


# ============================================================
# -- BezierMatrix --------------------------------------------
# ============================================================

struct BezierMatrix
    G      ::Matrix{Float64}
    axes   ::Vector{String}
    n      ::Int
    degree ::Int
    dim    ::Int
    M      ::Matrix{Float64}
    MG     ::Matrix{Float64}
end

function BezierMatrix(points::Matrix{Float64}, axes::Vector{String})
    n   = size(points, 1)
    deg = n - 1
    M   = derive_M(deg)
    BezierMatrix(points, axes, n, deg, length(axes), M, M * points)
end

function derive_M(deg::Int)::Matrix{Float64}
    n = deg
    M = zeros(Float64, n+1, n+1)
    for i in 0:n
        m = n - i
        for j in 0:m
            k = i + j
            M[n-k+1, i+1] = binomial(n,i) * binomial(m,j) * (-1)^j
        end
    end
    return M
end

function evaluate_all(bm::BezierMatrix, u_values::Vector{Float64})::Matrix{Float64}
    n    = bm.degree
    Uall = [u^(n-k) for u in u_values, k in 0:n]
    Uall * bm.MG
end


# ============================================================
# -- Formatting helpers --------------------------------------
# ============================================================

const WIDTH = 66
const SUPERSCRIPTS = Dict(0=>"", 1=>"", 2=>"²", 3=>"³", 4=>"⁴", 5=>"⁵")

ruler(char::Char='=') = string(char)^WIDTH

fmt(v::Float64) = abs(v - round(v)) < 1e-9 ? string(Int(round(v))) : @sprintf("%.4f", v)

function fmt_frac(v::Float64)
    r = rationalize(Int, v; tol=1e-9)
    abs(Float64(r) - v) < 1e-9 || return @sprintf("%.4f", v)
    isone(denominator(r)) ? string(numerator(r)) : string(r)
end

function fmt_poly(coeffs::Vector{Float64}, deg::Int)
    terms = Tuple{Float64,String}[]
    for (k, c) in enumerate(coeffs)
        k -= 1
        abs(c) < 1e-9 && continue
        power = deg - k
        u_s   = power == 0 ? "" : (power == 1 ? "u" : "u$(get(SUPERSCRIPTS, power, "^$power"))")
        c_s   = fmt(abs(c))
        c_s   = (c_s == "1" && !isempty(u_s)) ? "" : c_s
        push!(terms, (c, c_s * u_s))
    end
    isempty(terms) && return "0"
    parts = String[]
    for (idx, (c, t)) in enumerate(terms)
        push!(parts, idx == 1 ? (c < 0 ? "-$t" : t) : (c < 0 ? "- $t" : "+ $t"))
    end
    join(parts, "  ")
end

section(title::String) = (
    println();
    println("  $title");
    println("  " * "-"^(WIDTH-2))
)

# ── Plain-text table printer ─────────────────────────────────────────────────
# style :simple  → header + underline
# style :outline → box border

function print_table(
        rows       ::Vector{Vector{String}},
        headers    ::Vector{String};
        row_labels ::Union{Vector{String}, Nothing} = nothing,
        style      ::Symbol = :simple)

    has_rl = row_labels !== nothing
    widths = [length(h) for h in headers]
    for row in rows
        for (c, cell) in enumerate(row)
            widths[c] = max(widths[c], length(cell))
        end
    end
    rl_w = has_rl ? max(maximum(length.(row_labels)), 2) : 0

    function fmt_row(cells, label=nothing)
        parts = String[]
        has_rl && push!(parts, lpad(something(label, ""), rl_w))
        for (c, cell) in enumerate(cells)
            push!(parts, lpad(cell, widths[c]))
        end
        "  " * join(parts, "  ")
    end

    hdr      = fmt_row(headers, has_rl ? "" : nothing)
    ubar     = "  " * (has_rl ? " "^rl_w * "  " : "") * join(["-"^w for w in widths], "  ")
    inner_w  = sum(widths) + 2*(length(headers)-1) + (has_rl ? rl_w + 2 : 0)

    if style == :outline
        println("  ╒" * "═"^(inner_w+2) * "╕")
        println("  │ " * lstrip(hdr) * " │")
        println("  ╞" * "═"^(inner_w+2) * "╡")
        for (r, row) in enumerate(rows)
            println("  │ " * lstrip(fmt_row(row, has_rl ? row_labels[r] : nothing)) * " │")
        end
        println("  ╘" * "═"^(inner_w+2) * "╛")
    else
        println(hdr)
        println(ubar)
        for (r, row) in enumerate(rows)
            println(fmt_row(row, has_rl ? row_labels[r] : nothing))
        end
    end
end

function print_labeled_matrix(
        mat        ::Matrix{Float64},
        row_labels ::Vector{String},
        col_labels ::Vector{String};
        fmt_fn     = fmt)
    rows = [[fmt_fn(mat[r,c]) for c in 1:size(mat,2)] for r in 1:size(mat,1)]
    print_table(rows, col_labels; row_labels=row_labels)
end


# ============================================================
# -- Printer functions ---------------------------------------
# ============================================================

function print_title(bm::BezierMatrix, case_num::Int)
    names = Dict(2=>"Quadratic", 3=>"Cubic", 4=>"Quartic", 5=>"Quintic")
    dname = get(names, bm.degree, "Degree-$(bm.degree)")
    println()
    println(ruler())
    println("  Bezier Curve — Matrix Method  [P(u) = U · M · G]")
    println("  Case $case_num  |  $dname  |  $(bm.dim)D  |  $(bm.n) Control Points")
    println(ruler())
end

function print_G(bm::BezierMatrix)
    section("G  — Geometry Matrix (Control Points)")
    println()
    print_labeled_matrix(bm.G, ["P$(i-1)" for i in 1:bm.n], bm.axes)
end

function print_M(bm::BezierMatrix)
    section("M  — Basis Matrix (derived from Bernstein)")
    n = bm.degree
    println()
    print_labeled_matrix(bm.M, ["u^$(n-k)" for k in 0:n], ["P$(i-1)" for i in 1:bm.n])
end

function print_MG(bm::BezierMatrix)
    section("M · G  — Polynomial Coefficient Matrix")
    n = bm.degree
    println()
    print_labeled_matrix(bm.MG, ["u^$(n-k)" for k in 0:n], bm.axes)
end

function print_polynomials(bm::BezierMatrix)
    section("Resulting Polynomials")
    println()
    for (d, axis) in enumerate(bm.axes)
        println("  $(axis)(u)  =  $(fmt_poly(bm.MG[:,d], bm.degree))")
    end
end

function print_eval_table(bm::BezierMatrix, u_values::Vector{Float64})
    section("Evaluation Table  —  P(u) = U · (M · G)")
    pts  = evaluate_all(bm, u_values)
    rows = Vector{Vector{String}}()
    for (i, u) in enumerate(u_values)
        row = [fmt(u)]
        for d in 1:bm.dim
            v = pts[i, d]
            push!(row, "$(fmt_frac(v))  ($(@sprintf("%.4f", v)))")
        end
        push!(rows, row)
    end
    println()
    print_table(rows, vcat(["u"], ["$(ax)(u)" for ax in bm.axes]); style=:outline)
end

function print_all(bm::BezierMatrix, u_values::Vector{Float64}, case_num::Int)
    print_title(bm, case_num)
    print_G(bm)
    print_M(bm)
    print_MG(bm)
    print_polynomials(bm)
    print_eval_table(bm, u_values)
    println()
    println(ruler())
    println()
end


# ============================================================
# -- Entry Point ---------------------------------------------
# ============================================================

function main()
    data = CASE_DATA[CASE]
    bm   = BezierMatrix(data.points, data.axes)
    print_all(bm, data.u_values, CASE)
end

main()

File: bézier/python/04-bezier-matrix-straight.py
import numpy as np
from math import comb
from fractions import Fraction
from tabulate import tabulate


# ============================================================
# -- Switch  (only change this one line) ---------------------
# ============================================================

CASE = 2   # 1 = Quadratic 2D,  2 = Quartic 3D


# ============================================================
# -- Case Data  (do not change) ------------------------------
# ============================================================

CASE_DATA = {
    1: dict(
        points   = np.array([
            [1, 1],   # P0
            [3, 3],   # P1
            [5, 1],   # P2
        ], dtype=float),
        u_values = [0.00, 0.25, 0.50, 0.75, 1.00],
        axes     = ['x', 'y'],
    ),
    2: dict(
        points   = np.array([
            [0, 0, 1],   # P0
            [0, 4, 1],   # P1
            [4, 0, 1],   # P2
            [4, 4, 1],   # P3
            [5, 4, 1],   # P4
        ], dtype=float),
        u_values = [0.00, 0.25, 0.50, 0.75, 1.00],
        axes     = ['x', 'y', 'z'],
    ),
}


# ============================================================
# -- Bezier Matrix  (generic — do not change) ----------------
# ============================================================

class BezierMatrix:
    """
    Evaluates a Bezier curve of any degree via P(u) = U · M · G.

    M is derived automatically from the Bernstein definition.
    All evaluation is pure matrix multiplication — no loops
    over control points or basis functions at query time.

    M[k, i] = C(n,i) · C(n-i, k-i) · (-1)^(k-i)   for k >= i
            = 0                                       for k <  i

    Row k corresponds to power u^(n-k), so row 0 is u^n
    and the last row is the constant term u^0.
    This matches U = [u^n, u^(n-1), ..., u, 1].
    """

    def __init__(self, points: np.ndarray, axes: list):
        self.G      = points
        self.axes   = axes
        self.n      = len(points)
        self.degree = self.n - 1
        self.dim    = len(axes)
        self.M      = self._derive_M()
        self.MG     = self.M @ self.G   # pre-computed once

    def _derive_M(self) -> np.ndarray:
        n = self.degree
        M = np.zeros((n + 1, n + 1))
        for i in range(n + 1):
            m = n - i
            for j in range(m + 1):
                k         = i + j
                M[n-k, i] = comb(n, i) * comb(m, j) * ((-1) ** j)
        return M

    def U(self, u: float) -> np.ndarray:
        """Parameter row vector [u^n, u^(n-1), ..., u, 1]."""
        n = self.degree
        return np.array([[u ** (n - k) for k in range(n + 1)]])

    def evaluate_all(self, u_values: list) -> np.ndarray:
        """
        Evaluate at all u values in one batched matrix multiply.
        U_all shape (m, n+1)  ·  MG shape (n+1, dim)  →  (m, dim)
        """
        n     = self.degree
        U_all = np.array([
            [u ** (n - k) for k in range(n + 1)]
            for u in u_values
        ])
        return U_all @ self.MG


# ============================================================
# -- Console Printer  (generic — do not change) --------------
# ============================================================

class MatrixPrinter:
    """
    Prints G, M, M·G, polynomials, and evaluation table
    using tabulate for all matrix and table formatting.

    What tabulate replaces vs the previous version:
      _print_matrix  → tabulate() with 'simple' table format
      _print_table   → tabulate() with 'outline' table format
      _fmt / _fmt_fraction → kept only for polynomial strings
                             and cell values inside tabulate
    """

    WIDTH = 66

    def __init__(
        self,
        bm:       BezierMatrix,
        u_values: list,
        case_num: int,
    ):
        self.bm       = bm
        self.u_values = u_values
        self.case_num = case_num

    # -- Formatting helpers ----------------------------------

    def _ruler(self, char: str = '=') -> str:
        return char * self.WIDTH

    def _section(self, title: str):
        print()
        print(f'  {title}')
        print('  ' + '-' * (self.WIDTH - 2))

    def _fmt(self, v: float) -> str:
        """Integer if whole, else 4 decimal places."""
        if abs(v - round(v)) < 1e-9:
            return str(int(round(v)))
        return f'{v:.4f}'

    def _fmt_frac(self, v: float) -> str:
        """Exact fraction when possible, else 4 decimal places."""
        f = Fraction(v).limit_denominator(1024)
        if abs(float(f) - v) < 1e-9:
            return str(f) if f.denominator != 1 else str(f.numerator)
        return f'{v:.4f}'

    def _fmt_poly(self, coeffs: np.ndarray) -> str:
        """
        Build a polynomial string from a M·G column.
        coeffs[k]  =  coefficient of  u^(degree - k).
        """
        n   = self.bm.degree
        sup = {0:'', 1:'', 2:'²', 3:'³', 4:'⁴', 5:'⁵'}
        terms = []
        for k, c in enumerate(coeffs):
            if abs(c) < 1e-9:
                continue
            power = n - k
            u_s   = ('' if power == 0 else
                     'u' if power == 1 else
                     f'u{sup.get(power, f"^{power}")}')
            c_s   = self._fmt(abs(c))
            if c_s == '1' and u_s:
                c_s = ''
            terms.append((c, c_s + u_s))
        if not terms:
            return '0'
        parts = []
        for idx, (c, t) in enumerate(terms):
            parts.append(
                (f'-{t}' if c < 0 else t) if idx == 0
                else (f'- {t}' if c < 0 else f'+ {t}')
            )
        return '  '.join(parts)

    def _mat_to_table(
        self,
        mat:        np.ndarray,
        row_labels: list,
        col_labels: list,
        fmt_fn     = None,
    ) -> list:
        """
        Convert a numpy matrix to a list-of-lists suitable
        for tabulate, prepending row labels as the first column.
        fmt_fn defaults to self._fmt if not provided.
        """
        if fmt_fn is None:
            fmt_fn = self._fmt
        rows = []
        for r, rl in enumerate(row_labels):
            rows.append(
                [rl] + [fmt_fn(mat[r, c])
                        for c in range(mat.shape[1])]
            )
        return rows

    # -- Print sections --------------------------------------

    def print_all(self):
        self._print_title()
        self._print_G()
        self._print_M()
        self._print_MG()
        self._print_polynomials()
        self._print_table()
        print()
        print(self._ruler())
        print()

    def _print_title(self):
        names = {2:'Quadratic', 3:'Cubic',
                 4:'Quartic',   5:'Quintic'}
        dname = names.get(self.bm.degree, f'Degree-{self.bm.degree}')
        print()
        print(self._ruler())
        print(f'  Bezier Curve — Matrix Method  [P(u) = U · M · G]')
        print(f'  Case {self.case_num}  |  {dname}  |  '
              f'{self.bm.dim}D  |  {self.bm.n} Control Points')
        print(self._ruler())

    def _print_G(self):
        self._section('G  — Geometry Matrix (Control Points)')
        n = self.bm.degree
        rows    = self._mat_to_table(
            self.bm.G,
            row_labels = [f'P{i}' for i in range(self.bm.n)],
            col_labels = self.bm.axes,
        )
        print()
        print(tabulate(rows,
                       headers   = [''] + self.bm.axes,
                       tablefmt  = 'simple',
                       colalign  = ('left',)
                                   + ('right',) * self.bm.dim))

    def _print_M(self):
        self._section('M  — Basis Matrix (derived from Bernstein)')
        n       = self.bm.degree
        r_lbls  = [f'u^{n-k}' for k in range(self.bm.n)]
        c_lbls  = [f'P{i}'    for i in range(self.bm.n)]
        rows    = self._mat_to_table(self.bm.M, r_lbls, c_lbls)
        print()
        print(tabulate(rows,
                       headers  = [''] + c_lbls,
                       tablefmt = 'simple',
                       colalign = ('left',)
                                  + ('right',) * self.bm.n))

    def _print_MG(self):
        self._section('M · G  — Polynomial Coefficient Matrix')
        n      = self.bm.degree
        r_lbls = [f'u^{n-k}' for k in range(self.bm.n)]
        rows   = self._mat_to_table(self.bm.MG, r_lbls, self.bm.axes)
        print()
        print(tabulate(rows,
                       headers  = [''] + self.bm.axes,
                       tablefmt = 'simple',
                       colalign = ('left',)
                                  + ('right',) * self.bm.dim))

    def _print_polynomials(self):
        self._section('Resulting Polynomials')
        print()
        for d, axis in enumerate(self.bm.axes):
            poly = self._fmt_poly(self.bm.MG[:, d])
            print(f'  {axis}(u)  =  {poly}')

    def _print_table(self):
        self._section('Evaluation Table  —  P(u) = U · (M · G)')

        pts  = self.bm.evaluate_all(self.u_values)
        rows = []
        for u, pt in zip(self.u_values, pts):
            row = [self._fmt(u)]
            for v in pt:
                # Show exact fraction + decimal in same cell
                row.append(f'{self._fmt_frac(v)}  ({v:.4f})')
            rows.append(row)

        headers = ['u'] + [f'{ax}(u)' for ax in self.bm.axes]
        print()
        print(tabulate(rows,
                       headers  = headers,
                       tablefmt = 'outline',
                       colalign = ('right',) * (self.bm.dim + 1)))


# ============================================================
# -- Entry Point ---------------------------------------------
# ============================================================

def main() -> int:
    data = CASE_DATA[CASE]

    bm = BezierMatrix(
        points = data['points'],
        axes   = data['axes'],
    )

    printer = MatrixPrinter(
        bm       = bm,
        u_values = data['u_values'],
        case_num = CASE,
    )

    printer.print_all()
    return 0


if __name__ == '__main__':
    raise SystemExit(main())
